### ISSUE TITLE
print map metadata: title, author (version ?) on map loading screen

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#2114]
Creative people may want to be creative and make complex levelshots with text, symbols, maps or I don't know why, but we should not require mapper to bake map name and map author name on levelshot in order to get the map name and their own name being displayed at map load.

Also we better not require to go the image edition process each time the map change enough to get a new levelshot, taking a new levelshot without more edition should be enough.

The idea is to read map name and map author from `.arena` file and display it on map load screen with some RML.

For the map name, it's a good idea to have a reusable function that reads it from `.arena`, if missing reads it from bsp entity, if missing fallbacks on map bsp base name. This function would be reusable for https://github.com/Unvanquished/Unvanquished/issues/2113 .

We may even add license to .arena file.

**Comment by illwieckz** [comment]
This can be a good thread to also discuss the syntax and key names of that `.arena` file.

Right now, for a map named `castle`, the file is named `meta/castle/castle.arena`.

In tremulous it was named `scripts/castle.arena`

The content would be:

```
{
map "castle"
longname "Castle"
type "tremulous"
}
```

The long name can get spaces, special characters, color codes…

The `name` key is redundant, it exists for historical reasons, probably because some Q3-based games stored metadata for more than one map in one file. For example Smokin' Guns provide a file named `maps/sg_maps.cfg` that contains data for many maps:

```
{
	map	 	"br_backwater"
	author	        "Breli"
	longname	"Backwater"
	desc_line1	"----------------------------------"
	desc_line2	"Out in the nowhere ..."
	desc_line3	"----------------------------------"
	desc_line4	"BR Gametype"
}

{
	map		"dm_backwater"
	author	        "Breli"
	longname	"Backwater"
	desc_line1	"----------------------------------"
	desc_line2	"Out in the nowhere ..."
	desc_line3	"----------------------------------"
	desc_line4	"DM, TDM & RTP Gametypes"
}

{
	map		"dm_dawnfort"
	author	        "Breli"
	longname	"DawnFort"
	desc_line1	"----------------------------------"
	desc_line2	"Dawn has come to the fort ..."
	desc_line3	"----------------------------------"
	desc_line4	"DM, TDM & RTP Gametypes"
}
[…]
```

We don't support that variant, but that explains things…

About the `type` key, it is unused. Most maps set it to `tremulous` for historical reasons, some rare third-party maps set it to `unvanquished` (some maps by Matth features this). But as far as I know, this is not expected to be the game name, but the game type. In an arena game it would be deathmatch, clan arena, things like that.

It made sense at Tremulous time to set the game type to `tremulous` since Tremulous started as a Quake3 mod and then the game was run by Quake3 itself.

It would be good to be able to list multiple authors (and multiple licenses, maybe). A good way would be to repeat the keys, like this:

```
{
	map "atcshd"
	longname "Advanced Tremulous Combat Simulator"
	author "Jex"
	author "SuperPie"
	author "Ingar"
	license "CC-BY-SA-1.0"
	type "tremulous"
}
```

For licenses, SPDX short codes are fine.

For reference, the Xonotic game use a file named `maps/<basename>.mapinfo`, for example:

```
title Afterslime
description Medium sized indoor deathmatch map
author kuniu the frogg, t0uYK8Ne
cdtrack 12
has weapons
gametype dm
gametype tdm
gametype lms
gametype ca
gametype ft
gametype ka
gametype kh
gametype dom
```

The author key being a single string is a bit unfortunate, it means we can't look for a specific author in an easy way.

It's better if the sentence is built by the tool itself. Like “_Advanced Tremulous Combat Simulator by Jex, SuperPie and Ingar_”, with strings like `%s by %` and `% by % and %` being translatable.

The duplicate of keys is what does metadata formats like the ogg metadata format for music, or the matroska one for music/video, etc.

Example with some music files in my music folder:

```
title: Gavotte Pourlet
album: La nuit de la Saint Patrick
artist: Sonerien An Oriant
artist: Roñsed Mor
artist: Kevrenn Alré
artist: Kerlenn Pondi
artist: Bagadoù de Saint-Nazaire
artist: Cuivres du FIL
date: 2007-02-28
tracknumber: 05/14
discnumber: 01
```

```
title: Tricks Of The Light
album: Discovery
albumartist: Mike Oldfield
artist: Mike Oldfield
performer: Maggie Reilly
performer: Barry Palmer
date: 1984
genre: Progressive Rock
tracknumber: 4/13
discnumber: 01
```

**Comment by bmorel** [comment]
One thing. Defining formats is well and nice, but the .arena files are used since long, so we should support what currently exists.
From what I have seen, the file might be formatted roughly like this (pseudo EBNF description):

```
arena = "{" {white} KVlist {white} "}" ;
KVlist = { key white {white} value } ;
white = " " | "\t" | "\r" | "\n" ;
keyChar = "A" - "Z" | "a" - "z" | "0" - "9" | "-" | "_" ;
key = keyChar { keyChar } ;
valChar = ANY - "\"" ;
value = "\"" valChar { valChar } "\"" ;
```

**Comment by illwieckz** [comment]
As far as I know there is no `.arena` key for author name, neither for license, neither known way to handle multiple information for a given key. Maybe some games do this for type (on a single line?) but we don't support them anyway. So there is nothing that currently exists to support on those topics.

Hence the proposals for key names for metadata that don't exist yet and proposal for a format for multivalue keys and even print strings. That's assumed to be entirely new and can be integrated in the current format.

**Comment by bmorel** [comment]
imo the 1st thing to validate is the syntax. To have multiple values with the same key is not something difficult to do.

**Comment by illwieckz** [comment]
There are 1144 `.arena` files there: https://users.unvanquished.net/~illwieckz/dump/betaserv.tk/

**Comment by illwieckz** [comment]
The curly braces are also probably a reminiscence of a multi-map file format, like the `map` key.

**Comment by bmorel** [comment]
whatever the reason for the curly brace style, I'd rather have it supported to avoid problems. Nice for the 1K+ arena files. I'll see tomorrow if I have some time at work to setup a coco/R binary to check if they're conform to the EBNF I described earlier. Depending on my workload and willpower, I might have it running quickly.

**Comment by illwieckz** [comment]
If you script the extraction of them, make sure to exclude files from the #%@§! `__MACOSX` folder that may exist in some pk3.

**Comment by bmorel** [comment]
what are those exactly? I've seen them more than once, but no idea what is in those

**Comment by necessarily-equal** [comment]
If someone does the backend, I can do some RML & RCSS for the UI :smiley:

**Comment by cu-kai** [comment]
Closed by #2226
